File: Sorting/reorderLogFiles.cpp
// https://leetcode.com/problems/reorder-data-in-log-files//
// Version 1 - Stable partition then sort. All in place
// Time complexity      O(NlogN)
// Space complexity     O(N) if we're being strict but all mem allocated is temporary (partition, sort and std::string())
class Solution
{
public:
    vector<string> reorderLogFiles(vector<string> &logs)
    {
        // Partition so that all the digit logs are at then end in their relative order
        stable_partition(logs.begin(), logs.end(), isLetterLog);

        // Find the first digit log
        int n = logs.size();
        int firstDigLog = n - 1;
        for (int i = 0; i < n; ++i)
        {
            if (!isLetterLog(logs[i]))
            {
                firstDigLog = i;
                break;
            }
        }

        sort(logs.begin(), logs.begin() + firstDigLog,
             [](string &a, string &b) {
                 auto aFirstSpace = a.find(" "), bFirstSpace = b.find(" ");
                // Then sort logs before the index of that log:
                // either the contents of a are lexo less
                // or if the contents are equal
                // then we sort by ids
                 return (lexicographical_compare(a.begin() + aFirstSpace + 1, a.end(), b.begin() + bFirstSpace + 1, b.end()) ||
                         std::string(a.begin() + aFirstSpace, a.end()) == std::string(b.begin() + bFirstSpace, b.end()) &&
                             lexicographical_compare(a.begin(), a.begin() + aFirstSpace, b.begin(), b.begin() + bFirstSpace));
             });

        return logs;
    }

    static bool isLetterLog(string &log)
    {
        return !isdigit(log[log.find(" ") + 1]);
    }
};

// Version 2 - Found on LC. Split letter logs into ID and contents for easier sorting
class Solution
{
public:
    vector<string> reorderLogFiles(vector<string> &logs)
    {
        vector<string> digitLogs, ans;
        vector<pair<string, string>> letterLogs;
        for (string &s : logs)
        {
            int i = 0;
            while (s[i] != ' ')
                ++i;
            if (isalpha(s[i + 1]))
                letterLogs.emplace_back(s.substr(0, i), s.substr(i + 1));
            else
                digitLogs.push_back(s);
        }
        sort(letterLogs.begin(), letterLogs.end(), [&](auto &a, auto &b) {
            return a.second == b.second ? a.first < b.first : a.second < b.second;
        });
        for (auto &p : letterLogs)
            ans.push_back(p.first + " " + p.second);
        for (string &s : digitLogs)
            ans.push_back(s);
        return ans;
    }
};

File: Sorting/reorderLogFiles.py
# Version 1 - Using list.sort() with a key
# When we do key=lamdba x: ( f1(x), f2(x), f3(x) )
# We first try to sort in ascending order the keys produced from f1(x)
# If two elements from f1(x) producing the same key, we leave them in their existing order
# We then try to sort them based on the keys produced by f2(x)
# If again the keys are the same we try to sort by f3(x)
# And so on

# So for this problem the ordering is done as follows
# digitLogSortKey will return 0 for letter logs and (idx from original logs + 1) for a digit log
#   all the digit logs will be at then end of the array with their existing order and all the letter logs will be at the start of the array
#   because all the letter logs will have a key of 0 from this function, we move to the next function that produces a key
# getLogContents will return the content of the log as a key. By default Python sorts these lexicographically, as we wanted
#   if we have the same content we move to the next function that produces a key
# getLogId will return the Id of the log as a key. By default Python sorts these lexicographically, as we wanted

# Time complexity       O(n + nlogn)
# Space complexity       O(n)
class Solution:
    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        digit_log_ordering = {log : idx for idx, log in enumerate(logs) if self.is_digit_log(log)}
        logs.sort(key=lambda log: (self.digit_log_sort_key(log, digit_log_ordering), self.get_log_contents(log), self.get_log_id(log)))
        return logs

    @staticmethod
    def is_digit_log(log):
        return log[log.find(" ") + 1].isdigit()

    @staticmethod
    def get_log_contents(log):
        return log[log.find(" ") + 1:]

    @staticmethod
    def get_log_id(log):
        return log[:log.find(" ") ]

    @classmethod
    def digit_log_sort_key(cls, log, digit_log_ordering):
        if cls.is_digit_log(log):
            return digit_log_ordering[log] + 1
        return 0